### ISSUE TITLE
Precise documentation garantee

### ISSUE THREAD

**Issue by Sytten** [Nugine/simd#44]
Hi!
I have some questions regarding the statement `The runtime detection will be skipped if the fastest implementation is already available at compile-time` made in the documentation.

1. Does that mean that if I compile say on a 2024 machine and try to run that binary on a 2010 computer it might crash?
1. Is there a way to disable that behaviour / always force a runtime check?

Let me give you the context: I build my application on Github Actions on essentially unknown hardware and I then ship those binaries to a lot of users on various hardware (which can certainly be older than the Github Action machines). 
I cannot rely in any way, shape or form on the machine the binaries is compiled on.

Thanks :)

**Comment by Nugine** [comment]
You don't have to worry about it if your compiler settings are correct.

`rustc` generates instructions according to the target cpu settings. The default target cpu is usually very old for compatibility. `rustc --print target-cpus` shows the default.

If you enable more cpu features **explicitly** by setting `-C target-cpu=native` or `-C target-feature=+xxx+yyy+zzz`, then you are targeting more recent cpus. Older cpus cannot run the program which contains illegal instructions for them. In exchange, the program runs faster on recent cpus.

The cpu feature checks in simd crates
+ compile time: check the flags about the target cpu and target feature, nothing else
+ runtime: check the running environment including cpu and os

In conclusion, you will not be affected by illegal instructions problem unless `rustc` changes the default target cpu or you enable advanced cpu features explicitly.

BTW, if your host environment (GitHub Actions) has newer glibc than the target environment, you will probably come across the glibc version issue.

**Comment by Nugine** [comment]
For your second question, it is meaningless to disable the compile-time selection of fastest implementaion.

+ If the required cpu features have been enabled, then the advanced instructions are everywhere in the binary, not just one crate. Why not use them?
+ If the selection behavior is disabled, then the slower implementations will never be called, which causes unnecessary code bloat.

However, it might be meaningful to provide scalar-only mode for benchmark and smaller binary. Our scalar implementations are also fast!

**Comment by Sytten** [comment]
Thanks for the quick and comprehensive reply, much appreciated.
I understand better the guarantees now, might be worth adding this information in the documentation.
Yes I am aware of libc issues, we use `cross-rs` for this purpose.
